File: 73. Set Matrix Zeroes.py
class Solution:
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j] == 0:
                    for jj in range(len(matrix[i])):
                        if matrix[i][jj] != 0:
                            matrix[i][jj] = True
                    for ii in range(len(matrix)):
                        if matrix[ii][j] != 0:
                            matrix[ii][j] = True
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j] is True:
                    matrix[i][j] = 0


sol = Solution()
l1 = [[1, 1, 1],
      [1, 0, 1],
      [1, 1, 1]]
sol.setZeroes(l1)
print(l1)
l2 = [[0, 1, 2, 0],
      [3, 4, 5, 2],
      [1, 3, 1, 5]]
sol.setZeroes(l2)
print(l2)

File: _73_Set_Matrix_Zeroes.java
import java.util.Arrays;

class Solution {
    public void setZeroes(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return;
        int rows = matrix.length, cols = matrix[0].length;
        boolean firstColWithZero = false;
        for (int i=0; i < rows; ++i) {
            if (matrix[i][0] == 0) {
                firstColWithZero = true;
            }
            for (int j=1; j < cols; ++j) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        for (int i=1; i < rows; ++i) {
            for (int j=1; j < cols; ++j) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }
        
        if (matrix[0][0] == 0) {
            for (int j=0; j < cols; ++j) {
                matrix[0][j] = 0;
            }
        }
        if (firstColWithZero) {
            for (int i=0; i < rows; ++i) {
                matrix[i][0] = 0;
            }
        }
    }
}

public class _73_Set_Matrix_Zeroes {
    public static void main(String[] args) {
        Solution s = new Solution();
        int[][] a = {{0,1,2,0},{3,4,5,2},{1,3,1,5}};
        s.setZeroes(a);
        for (int[] r: a) System.out.println(Arrays.toString(r));
    }
}
